sefaria/model/text_manager.py

```diff
@@ -0,0 +1,142 @@
+import copy
+
+import django
+django.setup()
+from sefaria.model import *
+from sefaria.utils.hebrew import hebrew_term
+from typing import List
+
+class TextManager:
+    ALL = 'all'
+    BASE = 'base'
+    SOURCE = 'source'
+    TRANSLATION = 'translation'
+
+    def __init__(self, oref: Ref, versions_params: List[List[str]], fill_in_missing_segments=True):
+        self.versions_params = versions_params
+        self.oref = oref
+        self.fill_in_missing_segments = fill_in_missing_segments
+        self.handled_version_params = []
+        self.all_versions = self.oref.versionset()
+
+        fields = Version.optional_attrs + Version.required_attrs
+        fields.remove('chapter') # not metadata
+        self.return_obj = {
+            'versions': [],
+            'missings': [],
+            'available_langs': sorted({v.actualLanguage for v in self.all_versions}),
+            'available_versions': [{f: getattr(v, f, "") for f in fields} for v in self.all_versions]
+        }
+
+    def _append_version(self, version):
+        #TODO part of this function duplicate the functionality of Ref.versionlist(). maybe we should mvoe it to Version
+        fields = Version.optional_attrs + Version.required_attrs
+        for attr in ['chapter', 'title', 'language']:
+            fields.remove(attr)
+        version_details = {f: getattr(version, f, "") for f in fields}
+        text_range = TextRange(self.oref, version.actualLanguage, version.versionTitle, self.fill_in_missing_segments)
+
+        if self.fill_in_missing_segments:
+            # we need a new VersionSet of only the relevant versions for merging. copy should be better than calling for mongo
+            relevant_versions = copy.copy(self.all_versions)
+            relevant_versions.remove(lambda v: v.actualLanguage != version.actualLanguage)
```

Comment:
Do we test anywhere that the merge functionality in TextRange results in the same version being merged as the list of versions created here for the metadata of the merge?

Reply:
we have a test called test_fill_in_missing_segments. we didn't change the method of the merging itself which happens on VersionSet.merge.
If you have an idea for further testing, it will be great, please be more specific.